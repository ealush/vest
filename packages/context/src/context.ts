export default function createContext<T extends Record<string, unknown>>(
  init?: (ctxRef: T, parentContext: T | void) => T | null
): {
  run: <R>(ctxRef: T, fn: (context: T) => R) => R;
  bind: <Fn extends (...args: any[]) => any>(ctxRef: T, fn: Fn) => Fn;
  use: () => T | undefined;
} {
  const storage: { ctx?: T; ancestry: T[] } = { ancestry: [] };

  return {
    run,
    bind,
    use,
  };

  function run<R>(ctxRef: T, fn: (context: T) => R): R {
    const parentContext = use();

    const out = {
      ...(parentContext ? parentContext : {}),
      ...(init?.(ctxRef, parentContext) ?? ctxRef),
    };

    const ctx = set(Object.freeze(out));
    storage.ancestry.unshift(ctx);
    const res = fn(ctx);

    clear();
    return res;
  }

  function bind<Fn extends (...args: any[]) => any>(ctxRef: T, fn: Fn) {
    // @ts-ignore - this one's pretty hard to get
    const returnedFn: Fn = function (...runTimeArgs: Parameters<Fn>) {
      return run<ReturnType<Fn>>(ctxRef, function () {
        return fn(...runTimeArgs);
      });
    };

    return returnedFn;
  }

  function use() {
    return storage.ctx;
  }

  function set(value: T): T {
    return (storage.ctx = value);
  }

  function clear() {
    storage.ancestry.shift();
    set(storage.ancestry[0] ?? null);
  }
}
