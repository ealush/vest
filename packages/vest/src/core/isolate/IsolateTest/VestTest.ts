import { invariant, isPromise, optionalFunctionValue } from 'vest-utils';

import type { TIsolateTest } from 'IsolateTest';
import {
  TestAction,
  TestStateMachineAction,
  TestStatus,
  createTestStateMachine,
} from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { TFieldName, TGroupName } from 'SuiteResultTypes';

const TestStateMachine = createTestStateMachine();

export class VestTest {
  // Read

  static getData<
    F extends TFieldName = TFieldName,
    G extends TGroupName = TGroupName
  >(test: TIsolateTest<F, G>) {
    invariant(test.data);
    return test.data;
  }

  static warns(test: TIsolateTest): boolean {
    return VestTest.getData(test).severity === TestSeverity.Warning;
  }

  static isPending(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PENDING);
  }

  static isOmitted(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.OMITTED);
  }

  static isUntested(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.UNTESTED);
  }

  static isFailing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.FAILED);
  }

  static isCanceled(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.CANCELED);
  }

  static isSkipped(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.SKIPPED);
  }

  static isPassing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PASSING);
  }

  static isWarning(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.WARNING);
  }

  static hasFailures(test: TIsolateTest): boolean {
    return VestTest.isFailing(test) || VestTest.isWarning(test);
  }

  static isNonActionable(test: TIsolateTest): boolean {
    return (
      VestTest.isSkipped(test) ||
      VestTest.isOmitted(test) ||
      VestTest.isCanceled(test)
    );
  }

  static isTested(test: TIsolateTest): boolean {
    return VestTest.hasFailures(test) || VestTest.isPassing(test);
  }

  static awaitsResolution(test: TIsolateTest): boolean {
    // Is the test in a state where it can still be run, or complete running
    // and its final status is indeterminate?
    return (
      VestTest.isSkipped(test) ||
      VestTest.isUntested(test) ||
      VestTest.isPending(test)
    );
  }

  static isAsyncTest(test: TIsolateTest): boolean {
    return isPromise(VestTest.getData(test).asyncTest);
  }

  static statusEquals(test: TIsolateTest, status: TestStatus): boolean {
    return VestTest.getData(test).status === status;
  }

  // Mutate

  static setPending(test: TIsolateTest) {
    VestTest.setStatus(test, TestStatus.PENDING);
  }

  static fail(test: TIsolateTest): void {
    VestTest.setStatus(
      test,
      VestTest.warns(test) ? TestStatus.WARNING : TestStatus.FAILED
    );
  }

  static pass(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.PASSING);
  }

  static warn(test: TIsolateTest): void {
    VestTest.setData(test, current => ({
      ...current,
      severity: TestSeverity.Warning,
    }));
  }

  static setData(
    test: TIsolateTest,
    setter:
      | ((current: TIsolateTest['data']) => TIsolateTest['data'])
      | TIsolateTest['data']
  ): void {
    test.data = optionalFunctionValue(setter, VestTest.getData(test));
  }

  static skip(test: TIsolateTest, force?: boolean): void {
    // Without this force flag, the test will be marked as skipped even if it is pending.
    // This means that it will not be counted in "allIncomplete" and its done callbacks
    // will not be called, or will be called prematurely.
    // What this mostly say is that when we have a pending test for one field, and we then
    // start typing in a different field - the pending test will be canceled, which
    // is usually an unwanted behavior.
    // The only scenario in which we DO want to cancel the async test regardless
    // is when we specifically skip a test with `skipWhen`, which is handled by the
    // "force" boolean flag.
    // I am not a fan of this flag, but it gets the job done.
    VestTest.setStatus(test, TestStatus.SKIPPED, force);
  }

  static cancel(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.CANCELED);
    VestTest.getData(test).abortController.abort(TestStatus.CANCELED);
  }

  static omit(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.OMITTED);
  }

  static reset(test: TIsolateTest): void {
    VestTest.setStatus(test, TestAction.RESET);
  }

  static setStatus(
    test: TIsolateTest,
    status: TestStateMachineAction,
    payload?: any
  ): void {
    VestTest.setData(test, current => ({
      ...current,
      status: TestStateMachine.staticTransition(
        current.status,
        status,
        payload
      ),
    }));
  }
}
