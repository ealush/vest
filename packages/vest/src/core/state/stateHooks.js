import ctx from 'ctx';

const getStateRef = () => ctx.use().stateRef;

export function usePending() {
  return getStateRef().pending();
}
export function useSuiteId() {
  return getStateRef().suiteId();
}
export function useTestCallbacks() {
  return getStateRef().testCallbacks();
}
export function useTestObjects() {
  return getStateRef().testObjects();
}
export function useSkippedTests() {
  return getStateRef().skippedTests();
}
export function useOptionalFields() {
  return getStateRef().optionalFields();
}
