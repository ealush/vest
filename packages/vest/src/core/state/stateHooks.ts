import ctx from 'ctx';

export function useCarryOverTests() {
  return useStateRef().carryOverTests();
}
export function usePending() {
  return useStateRef().pending();
}
export function useLagging() {
  return useStateRef().lagging();
}
export function useSuiteId() {
  return useStateRef().suiteId();
}
export function useTestCallbacks() {
  return useStateRef().testCallbacks();
}
export function useTestObjects() {
  return useStateRef().testObjects();
}
export function useSkippedTests() {
  return useStateRef().skippedTests();
}
export function useOptionalFields() {
  return useStateRef().optionalFields();
}
export function useStateRef() {
  return ctx.use()!.stateRef!;
}
