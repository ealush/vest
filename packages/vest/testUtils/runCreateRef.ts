import createState from 'vast';

import createStateRef from 'createStateRef';

export default (
  state?: ReturnType<typeof createState>
): ReturnType<typeof createStateRef> =>
  createStateRef(state ? state : createState(), {
    suiteId: '1000',
  });
