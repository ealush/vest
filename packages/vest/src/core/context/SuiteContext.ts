import { createCascade } from 'context';
import { assign, TinyState, tinyState, cache, CacheApi } from 'vest-utils';
import { DynamicValue } from 'vest-utils/src/utilityTypes';

import { IsolateTest } from 'IsolateTest';
import { Modes } from 'Modes';

export const SuiteContext = createCascade<CTXType>((ctxRef, parentContext) => {
  if (parentContext) {
    return null;
  }

  return assign(
    {
      inclusion: {},
      mode: tinyState.createTinyState<Modes>(Modes.EAGER),
      suiteParams: [],
      testMemoCache,
    },
    ctxRef
  );
});

type CTXType = {
  inclusion: Record<string, DynamicValue<boolean>>;
  mode: TinyState<Modes>;
  suiteParams: any[];
  testMemoCache: CacheApi<IsolateTest>;
  currentTest?: IsolateTest;
  groupName?: string;
  skipped?: boolean;
  omitted?: boolean;
};

export function useCurrentTest(msg?: string) {
  return SuiteContext.useX(msg).currentTest;
}

export function useGroupName() {
  return SuiteContext.useX().groupName;
}

export function useInclusion() {
  return SuiteContext.useX().inclusion;
}

export function useMode() {
  return SuiteContext.useX().mode();
}

export function useSkipped() {
  return SuiteContext.useX().skipped ?? false;
}

export function useOmitted() {
  return SuiteContext.useX().omitted ?? false;
}

const testMemoCache = cache<IsolateTest>(10);

export function useTestMemoCache() {
  return SuiteContext.useX().testMemoCache;
}

export function useSuiteParams() {
  return SuiteContext.useX().suiteParams;
}
