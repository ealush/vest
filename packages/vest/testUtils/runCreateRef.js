import state from 'state';
import usePending from 'usePending';
import useSuiteId from 'useSuiteId';
import useTestCallbacks from 'useTestCallbacks';
import useTestObjects from 'useTestObjects';

export default () =>
  state.createRef({
    usePending,
    useSuiteId: [useSuiteId, [1000, 'suite_name']],
    useTestCallbacks,
    useTestObjects,
  });
