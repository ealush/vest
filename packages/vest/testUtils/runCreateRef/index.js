import useTestCallbacks from '../../src/core/produce/useTestCallbacks';
import state from '../../src/core/state';
import useSuiteId from '../../src/core/suite/useSuiteId';
import usePending from '../../src/core/test/lib/pending/usePending';
import useTestObjects from '../../src/core/test/useTestObjects';

export default () =>
  state.createRef({
    usePending,
    useSuiteId: [useSuiteId, [1000, 'suite_name']],
    useTestCallbacks,
    useTestObjects,
  });
