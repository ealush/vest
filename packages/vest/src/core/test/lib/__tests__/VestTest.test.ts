import runCreateRef from '../../../../../testUtils/runCreateRef';

import VestTest from 'VestTest';
import addTestToState from 'addTestToState';
import context from 'ctx';
import { setPending } from 'pending';
import { usePending, useTestObjects, useLagging } from 'stateHooks';


const fieldName = 'unicycle';
const message = 'I am Root.';

let stateRef;

describe('VestTest', () => {
  let testObject;

  beforeEach(() => {
    testObject = new VestTest(fieldName, jest.fn(), {
      message,
    });
  });

  test('TestObject constructor', () => {
    expect(testObject).toMatchSnapshot();
  });

  it('Should have a unique id', () => {
    Array.from(
      { length: 100 },
      () => new VestTest(fieldName, jest.fn(), { message })
    ).reduce((existing, { id }) => {
      expect(existing[id]).toBeUndefined();
      existing[id] = true;
      return existing;
    }, {});
  });

  describe('testObject.warn', () => {
    it('Should set `.isWarning` to true', () => {
      expect(testObject.isWarning).toBe(false);
      testObject.warn();
      expect(testObject.isWarning).toBe(true);
      expect(testObject).toMatchSnapshot();
    });
  });

  describe('testObject.fail', () => {
    beforeEach(() => {
      jest.resetModules();

      const VestTest = require('VestTest').default;
      testObject = new VestTest(fieldName, jest.fn(), { message });
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    it('Should set this.failed to true', () => {
      expect(testObject.failed).toBe(false);
      testObject.fail();
      expect(testObject.failed).toBe(true);
    });
  });

  describe('testobject.valueOf', () => {
    test('When `failed` is false', () => {
      expect(testObject.failed).toBe(false);
      expect(testObject.valueOf()).toBe(true);
    });

    test('When `failed` is true', () => {
      testObject.failed = true;
      expect(testObject.valueOf()).toBe(false);
    });
  });

  describe('testObject.cancel', () => {
    const getCtx = () => ({ stateRef });

    beforeEach(() => {
      stateRef = runCreateRef();

      context.run({ stateRef }, () => {
        addTestToState(testObject);
      });
    });

    it.withContext(
      'Should remove a testObject from the state',
      () => {
        const [testObjects] = useTestObjects();
        expect(testObjects).toEqual(expect.arrayContaining([testObject]));
        testObject.cancel();
        expect(testObjects).toEqual(expect.not.arrayContaining([testObject]));
      },
      getCtx
    );

    it.withContext('Should remove a testObject from the pending state', () => {
      setPending(testObject);
      {
        const [pending] = usePending();
        expect(pending).toEqual(expect.arrayContaining([testObject]));
      }
      testObject.cancel();
      {
        const [pending] = usePending();
        expect(pending).toEqual(expect.not.arrayContaining([testObject]));
      }
    });

    it.withContext('Should remove a testObject from the lagging state', () => {
      const [, setLagging] = useLagging();
      setLagging(() => [testObject]);
      {
        const [lagging] = useLagging();
        expect(lagging).toEqual(expect.arrayContaining([testObject]));
      }
      testObject.cancel();
      {
        const [lagging] = useLagging();
        expect(lagging).toEqual(expect.not.arrayContaining([testObject]));
      }
    });
  });
});
