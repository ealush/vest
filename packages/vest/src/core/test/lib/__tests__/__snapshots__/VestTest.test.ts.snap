// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VestTest TestObject constructor 1`] = `
VestTest {
  "canceled": false,
  "failed": false,
  "fieldName": "unicycle",
  "id": "0",
  "isWarning": false,
  "message": "I am Root.",
  "testFn": [MockFunction],
}
`;

exports[`VestTest testObject.warn Should set \`.isWarning\` to true 1`] = `
VestTest {
  "canceled": false,
  "failed": false,
  "fieldName": "unicycle",
  "id": "102",
  "isWarning": true,
  "message": "I am Root.",
  "testFn": [MockFunction],
}
`;
